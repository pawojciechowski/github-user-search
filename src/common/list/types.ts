export interface ListItem {
  id: string;
  text: string;
  href?: string;
}
