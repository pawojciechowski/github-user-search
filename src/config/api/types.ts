// TODO: Verify types

export interface GithubUserResponse {
  login: string;
  name: string | null;
  bio: string | null;
  avatar_url: string;
}

export interface GithubReposResponseRepository {
  id: number;
  name: string;
  html_url: string;
  stargazers_count: number;
}

export interface GithubReposResponse extends Array<GithubReposResponseRepository> {}
