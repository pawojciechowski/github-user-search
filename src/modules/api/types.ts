export type APIUrl = string;

export interface IAPIService {
  get<T>(url: APIUrl, params?: { [key: string]: string }): Promise<T>;
}

export interface APIServiceOptions {
  headers?: { [key: string]: string };
  mockResponse?: { [key: string]: any };
  logRequests?: boolean;
}
