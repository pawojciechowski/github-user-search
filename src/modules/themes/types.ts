import { themeNames } from './themes';

export type Color = string;
export type FontSize = string;
export type FontWeight = string;
export type FontFamily = string;
export type LineHeight = string;
export type LetterSpacing = string;

interface FontSettings {
  fontFamily?: FontFamily,
  fontSize?: FontSize,
  fontWeight?: FontWeight,
  lineHeight?: LineHeight,
  letterSpacing?: LetterSpacing,
  color?: Color
}

interface InputStyles {
  placeholder?: {
    text: FontSettings
  },
  background: Color,
  text: FontSettings
}

interface ButtonStyles {
  background: Color,
  text: FontSettings
}

interface TypographySettings {
  base: FontSettings,
  title: FontSettings
}

export interface Theme {
  background: Color,
  input: InputStyles,
  button: ButtonStyles,
  spacer: number,
  typography: TypographySettings,
}

export type ThemeName = typeof themeNames[number];
