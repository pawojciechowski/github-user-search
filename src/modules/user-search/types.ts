export interface GithubUserResponse {
  login: string,
  name: string,
  bio: string | null,
  avatar_url: string
}

export interface GithubReposResponseRepo {
  id: number,
  name: string,
  html_url: string,
  stargazers_count: number
}

export interface GithubReposResponse extends Array<GithubReposResponseRepo> {}
