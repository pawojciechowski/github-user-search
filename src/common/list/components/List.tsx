import React from 'react';
import styled, { css } from 'styled-components';
import Title from 'modules/themes/components/Title';
import { Theme } from 'modules/themes/types';
import { ListItem } from '../types';
import Skeleton from 'react-loading-skeleton';

const ListContainer = styled.div`
  ul {
    list-style: none;
    padding: 0;
    margin: 0;
    display: flex;
    flex-direction: column;
  }

  li {
    display: block;
    margin-bottom: ${({ theme }: { theme: Theme }) => `${theme.spacer}px`};
  }
`;

const listItemStyles = css`
  display: block;
  padding: ${({ theme }: { theme: Theme }) => `${theme.spacer}px`};
  background: ${({ theme }: { theme: Theme }) => `${theme.foreground}`};
  box-shadow: ${({ theme }: { theme: Theme }) => `${theme.containerShadow}`};
  border-radius: 12px;
  line-height: 16px;
  color: ${({ theme }: { theme: Theme }) => `${theme.secondary}`};
`;

const ListItemContent = styled.div`
  ${listItemStyles}
`;

const ListItemContentHref = styled.a`
  ${listItemStyles}
  text-decoration: none;
`;

interface ListProps {
  title: string;
  items: ListItem[];
  loading?: boolean;
}

export function List({ title, items, loading }: ListProps) {
  return (
    <ListContainer>
      <Title as="h2">{title}</Title>
      <ul>
        {!loading
          ? items.map((item) => {
              return (
                <li key={item.id}>
                  {item.href ? (
                    <ListItemContentHref rel="noopener nofollow noreferer" target="_blank" href={item.href}>
                      {item.text}
                    </ListItemContentHref>
                  ) : (
                    <ListItemContent>{item.text}</ListItemContent>
                  )}
                </li>
              );
            })
          : [...(Array(3).keys() as any)].map((i) => {
              return (
                <li key={i}>
                  <Skeleton height={48} />
                </li>
              );
            })}
      </ul>
    </ListContainer>
  );
}

export default List;
