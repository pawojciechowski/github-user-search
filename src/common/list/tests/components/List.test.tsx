import React from 'react';
import { render } from '@testing-library/react';
import List, { ListContainer, ListItemContent } from '../../components/List';
import { ListItem } from 'common/list/types';
import { renderWithThemeProvider, testTheme } from 'config/tests/utils';

const testItemsWithHrefs: ListItem[] = [
  {
    id: '1',
    text: 'test1',
    href: 'http://localhost1',
  },
  {
    id: '2',
    text: 'test2',
    href: 'http://localhost2',
  },
];

const testItemsWithoutHrefs: ListItem[] = [
  {
    id: '1',
    text: 'test1',
  },
  {
    id: '2',
    text: 'test2',
  },
];

const testTitle = 'testTitle';

describe('<ListContainer />', () => {
  it('matches snapshot', () => {
    const { container } = render(<ListContainer theme={testTheme} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('<ListItemContent />', () => {
  it('matches snapshot', () => {
    const { container } = render(<ListItemContent theme={testTheme} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('<ListItemContentHref />', () => {
  it('matches snapshot', () => {
    const { container } = render(<ListItemContent theme={testTheme} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('<List />', () => {
  it('renders proper number of items', () => {
    const { container: container1 } = renderWithThemeProvider(<List items={testItemsWithHrefs} title={testTitle} />);
    expect(container1.querySelector('ul')?.children.length).toBe(2);

    const { container: container2 } = renderWithThemeProvider(<List items={[]} title={testTitle} />);
    expect(container2.querySelector('ul')?.children.length).toBe(0);
  });

  it('renders title text', () => {
    const { queryByText } = renderWithThemeProvider(<List items={testItemsWithHrefs} title={testTitle} />);
    expect(queryByText(testTitle)).toBeTruthy();
  });

  it('renders in anchor element when href is passed', () => {
    const { container } = renderWithThemeProvider(<List items={testItemsWithHrefs} title={testTitle} />);
    expect(container.querySelectorAll('li > a').length).toBe(2);
  });

  it('renders in div element when href is not passed', () => {
    const { container } = renderWithThemeProvider(<List items={testItemsWithoutHrefs} title={testTitle} />);
    expect(container.querySelectorAll('li > div').length).toBe(2);
  });

  it('matches snapshot', () => {
    const { container } = renderWithThemeProvider(<List items={testItemsWithHrefs} title={testTitle} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot when loading true', () => {
    const { container } = renderWithThemeProvider(<List items={testItemsWithHrefs} title={testTitle} loading={true} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
