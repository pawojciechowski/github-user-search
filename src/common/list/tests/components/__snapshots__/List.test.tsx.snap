// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> matches snapshot 1`] = `
.c1 {
  font-size: test;
  font-weight: test;
  line-height: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  margin-top: 0;
  margin-bottom: 1px;
  padding: 0;
}

.c0 ul {
  list-style: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 li {
  display: block;
  margin-bottom: 1px;
}

.c2 {
  display: block;
  padding: 1px;
  background: test;
  box-shadow: test;
  border-radius: 12px;
  line-height: 16px;
  color: test;
  -webkit-text-decoration: none;
  text-decoration: none;
}

<div
  class="c0"
>
  <h2
    class="c1"
  >
    testTitle
  </h2>
  <ul>
    <li>
      <a
        class="c2"
        href="http://localhost1"
        rel="noopener nofollow noreferer"
        target="_blank"
      >
        test1
      </a>
    </li>
    <li>
      <a
        class="c2"
        href="http://localhost2"
        rel="noopener nofollow noreferer"
        target="_blank"
      >
        test2
      </a>
    </li>
  </ul>
</div>
`;

exports[`<ListContainer /> matches snapshot 1`] = `
.c0 ul {
  list-style: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 li {
  display: block;
  margin-bottom: 1px;
}

<div
  class="c0"
/>
`;

exports[`<ListItemContent /> matches snapshot 1`] = `
.c0 {
  display: block;
  padding: 1px;
  background: test;
  box-shadow: test;
  border-radius: 12px;
  line-height: 16px;
  color: test;
}

<div
  class="c0"
/>
`;

exports[`<ListItemContentHref /> matches snapshot 1`] = `
.c0 {
  display: block;
  padding: 1px;
  background: test;
  box-shadow: test;
  border-radius: 12px;
  line-height: 16px;
  color: test;
}

<div
  class="c0"
/>
`;
